core/jvm/src/main/scala/scalaz/zio/RTS.scala

```diff
@@ -255,15 +252,9 @@ private object RTS {
           case a: IO.Attempt[_, _, _, _] =>
             errorHandler = a.err.asInstanceOf[Any => IO[Any, Any]]
           case f0: Finalizer[_] =>
-            val f = f0.asInstanceOf[Finalizer[E]]
-
-            // Lazy initialization of body:
-            if (body eq null) body = ExitResult.Failed(err)
-
-            val currentFinalizer: IO[E2, List[Throwable]] = f.finalizer(body).run.map(collectDefect)
-
-            if (finalizer eq null) finalizer = currentFinalizer
-            else finalizer = finalizer.zipWith(currentFinalizer)(_ ++ _)
+            val f: IO[Void, List[Throwable]] = f0.finalizer.run.map(collectDefect)
```

Comment:
If you want, you can make this faster with:

```scala
case f0 if f0.getClass == Finalizer.class` =>
```

It will be a little faster than an `instanceOf` check. Same for `IO.Attempt`.

Reply:
[someone run a benchmark](https://stackoverflow.com/questions/103564/the-performance-impact-of-using-instanceof-in-java/26514984#26514984) that would suggest that `_.isInstanceOf[Finalizer[_]]` is slightly faster (but basically same perf) than `_.getClass eq classOf[Finalizer[_]]` .

Reply:
All right, then let's not change it. 👍 

Must be due to performance optimizations since the reverse used to be true...